File: src/primitives/Hash/Hash.swift
import Foundation
import CVoltaire

/// 32-byte hash value (Keccak-256, etc.)
public struct Hash: Equatable, Hashable, CustomStringConvertible, Sendable {
    /// Raw C struct
    internal var raw: PrimitivesHash

    /// Zero hash
    public static let zero = Hash()

    /// Create zero hash
    public init() {
        self.raw = PrimitivesHash()
    }

    /// Create from hex string (with or without 0x prefix)
    public init(hex: String) throws {
        var hash = PrimitivesHash()
        let result = hex.withCString { ptr in
            primitives_hash_from_hex(ptr, &hash)
        }
        try checkResult(result)
        self.raw = hash
    }

    /// Create from raw bytes
    public init(bytes: [UInt8]) throws {
        guard bytes.count == 32 else {
            throw VoltaireError.invalidLength
        }
        var hash = PrimitivesHash()
        withUnsafeMutableBytes(of: &hash.bytes) { dest in
            bytes.withUnsafeBytes { src in
                dest.copyMemory(from: src)
            }
        }
        self.raw = hash
    }

    /// Create from C API type
    internal init(cHash: PrimitivesHash) {
        self.raw = cHash
    }

    /// Convert to hex string (lowercase, with 0x prefix)
    public var hex: String {
        var hash = raw
        var buf = [CChar](repeating: 0, count: 67)
        primitives_hash_to_hex(&hash, &buf)
        return String(cString: buf)
    }

    /// Convert to byte array
    public var bytes: [UInt8] {
        withUnsafeBytes(of: raw.bytes) { Array($0) }
    }

    // MARK: - CustomStringConvertible

    public var description: String {
        hex
    }

    // MARK: - Equatable (constant-time)

    public static func == (lhs: Hash, rhs: Hash) -> Bool {
        var lhsRaw = lhs.raw
        var rhsRaw = rhs.raw
        return primitives_hash_equals(&lhsRaw, &rhsRaw)
    }

    // MARK: - Hashable

    public func hash(into hasher: inout Hasher) {
        withUnsafeBytes(of: raw.bytes) { hasher.combine(bytes: $0) }
    }
}

File: src/primitives/Hash/Hash.kt
package com.voltaire

import java.nio.charset.StandardCharsets

/** 32-byte hash value (Keccak-256, etc.) */
class Hash internal constructor(internal val raw: PrimitivesHash) {

    companion object {
        /** Zero hash */
        @JvmStatic
        val ZERO: Hash = Hash(PrimitivesHash())

        /** Create from hex string (with or without 0x prefix) */
        @JvmStatic
        @Throws(VoltaireError::class)
        fun fromHex(hex: String): Hash {
            val hash = PrimitivesHash()
            checkResult(VoltaireLib.INSTANCE.primitives_hash_from_hex(hex, hash))
            return Hash(hash)
        }

        /** Create from raw bytes */
        @JvmStatic
        @Throws(VoltaireError::class)
        fun fromBytes(bytes: ByteArray): Hash {
            if (bytes.size != 32) throw VoltaireError.InvalidLength
            val hash = PrimitivesHash()
            System.arraycopy(bytes, 0, hash.bytes, 0, 32)
            return Hash(hash)
        }
    }

    /** Hex string with 0x prefix */
    val hex: String
        get() {
            val buf = ByteArray(67)
            val rc = VoltaireLib.INSTANCE.primitives_hash_to_hex(raw, buf)
            require(rc >= 0) { "primitives_hash_to_hex failed: $rc" }
            return String(buf, 0, 66, StandardCharsets.UTF_8)
        }

    /** Raw bytes */
    val bytes: ByteArray
        get() = raw.bytes.copyOf()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Hash) return false
        return VoltaireLib.INSTANCE.primitives_hash_equals(raw, other.raw) != 0.toByte()
    }

    override fun hashCode(): Int = raw.bytes.contentHashCode()

    override fun toString(): String = hex
}
